gpu_direct_rdma_access.c

```diff
@@ -838,6 +866,66 @@ struct rdma_device *rdma_open_device_server(struct sockaddr *addr)
     return NULL;
 }
 
+int rdma_reset_server_device(struct rdma_task_attr *attr)
+{
+	struct rdma_device *device = attr->local_buf_rdma->rdma_dev;
+	if (is_server(device)) {
+		fprintf(stderr, "Method \"rdma_reset_server_device()\" could be executed only by server side!\n");
+		return EOPNOTSUPP;
+	}
+	struct ibv_qp_attr      qp_attr;
+	enum ibv_qp_attr_mask   attr_mask;
+	memset(&qp_attr, 0, sizeof qp_attr);
+	
+	/* - - - - - - - Modify QP to ERR - - - - - - - */
+	qp_attr.qp_state = IBV_QPS_ERR;
+	attr_mask = IBV_QP_STATE;
+	DEBUG_LOG("ibv_modify_qp(qp = %p, qp_attr.qp_state = %d, attr_mask = 0x%x)\n",
+                      device->qp, qp_attr.qp_state, attr_mask);
+	if (ibv_modify_qp(device->qp, &qp_attr, attr_mask)) {
+		fprintf(stderr, "Failed to modify QP to ERR\n");
+		return 1;
+	}
+	
+	/* - - - - - - - FLUSH WORK COMPLETIONS - - - - - - - */
+	attr->flags |= RDMA_TASK_ATTR_ZERO_BYTE_MSG;
+	attr->wr_id = ZERO_BYTE_MSG_WR_ID;
+	rdma_submit_task(attr);
+
+	DEBUG_LOG_FAST_PATH("Polling ZERO_BYTE_MSG completion\n");
+	struct rdma_completion_event rdma_comp_ev[16];
```

Comment:
16?
we should be ready to get SEND_Q_DEPTH completions.
you can break this into bursts of 16 if it helps

Reply:
it's internal restriction, so I call it with max  size inside do while:
```
num_entries = COMP_ARRAY_SIZE; /* We don't returne more than 16 entries,
 If user needs more, he can call rdma_poll_completions again */
```

Reply:
so, you should use the COMP_ARRAY_SIZE and not 16 to specific it's the same value.

BUT, wrong.
the CQ_DEPTH is 640, and although you should limit the rdma_poll_comp(num<=16), you're moving in the rdma_comp_ev[] forward until ALL CQE's are poll'ed, which might be 640 at worse case.

Reply:
it seems that rdma_poll_completions() should be fixed so it returns ONLY events which have the real app_wr_id, and skip the internal posted WQE's that failed but should not be reported.

this is besides the getting ready for CQ_DEPTH completions

Reply:
> it seems that rdma_poll_completions() should be fixed so it returns ONLY events which have the real app_wr_id, and skip the internal posted WQE's that failed but should not be reported.
> 
> this is besides the getting ready for CQ_DEPTH completions

it already fixed by active_flag inside wr_id_reported, that's means that poll_compl will return only not failed real app_wr_id's (only 1 real wr_id per submit task).

Reply:
> so, you should use the COMP_ARRAY_SIZE and not 16 to specific it's the same value.
> 
> BUT, wrong.
> the CQ_DEPTH is 640, and although you should limit the rdma_poll_comp(num<=16), you're moving in the rdma_comp_ev[] forward until ALL CQE's are poll'ed, which might be 640 at worse case.

That's right, we can call **rdma_poll_completions** function with <=16 entries step, I agree that it's better to use defined variable.
I'm not sure that for **rdma_comp_ev[]** such size is necessary, because this call in general return only one reported entry (zero byte msg wr)